File: backend/_routes/def/invite.ts
import { createRoute, z } from "npm:@hono/zod-openapi";
import { Invite, InviteCode, InvitePatch, RoomId } from "../types.ts";

export const InviteCreate = createRoute({
	method: "post",
	path: "/api/v1/invites",
	summary: "Invite create",
	tags: ["invite"],
	request: {
		body: {
			content: {
				"application/json": {
					schema: InvitePatch,
				},
			},
		},
	},
	responses: {
		201: {
			description: "created",
			content: {
				"application/json": {
					schema: Invite,
				},
			},
		},
	},
});

export const InviteDelete = createRoute({
	method: "delete",
	path: "/api/v1/invites/{invite_code}",
	summary: "Invite delete",
	tags: ["invite"],
	request: {
		params: z.object({
			invite_code: InviteCode,
		}),
		body: {
			content: {
				"application/json": {
					schema: InvitePatch,
				},
			},
		},
	},
	responses: {
		204: {
			description: "success",
		},
	},
});

export const InviteResolve = createRoute({
	method: "get",
	path: "/api/v1/invites/{invite_code}",
	summary: "Invite resolve",
	tags: ["invite"],
	request: {
		params: z.object({
			invite_code: InviteCode,
		}),
	},
	responses: {
		200: {
			description: "success",
			content: {
				"application/json": {
					schema: Invite,
				},
			},
		},
	},
});

export const InviteUse = createRoute({
	method: "post",
	path: "/api/v1/invites/{invite_code}",
	summary: "Invite use",
	tags: ["invite"],
	request: {
		params: z.object({
			invite_code: InviteCode,
		}),
	},
	responses: {
		200: {
			description: "success",
			content: {
				// TODO: schema
				// "application/json": {
				//   schema: ,
				// }
			},
		},
	},
});

export const InviteRoomList = createRoute({
	method: "get",
	path: "/api/v1/rooms/{room_id}/invites",
	summary: "Invite room list",
	tags: ["invite"],
	params: z.object({
		room_id: RoomId,
	}),
	responses: {
		200: {
			description: "success",
			content: {
				"application/json": {
					schema: z.object({
						invites: Invite.array(),
					}),
				},
			},
		},
	},
});

File: crate-common/src/v1/routes/invite.rs
use lamprey_macros::endpoint;

/// Invite delete
#[endpoint(
    delete,
    path = "/invite/{invite_code}",
    tags = ["invite"],
    scopes = [Full],
    permissions_optional = [InviteManage],
    audit_log_events = ["InviteDelete"],
    response(NO_CONTENT, description = "success"),
)]
pub mod invite_delete {
    use crate::v1::types::InviteCode;

    pub struct Request {
        #[path]
        pub invite_code: InviteCode,
    }

    pub struct Response {}
}

/// Invite resolve
#[endpoint(
    get,
    path = "/invite/{invite_code}",
    tags = ["invite"],
    scopes = [Full],
    response(OK, body = Invite, description = "success"),
    response(OK, body = InviteWithMetadata, description = "success with metadata"),
)]
pub mod invite_resolve {
    use crate::v1::types::{Invite, InviteCode, InviteWithMetadata};

    pub struct Request {
        #[path]
        pub invite_code: InviteCode,
    }

    pub struct Response {
        #[json]
        pub invite: Invite,
    }
}

/// Invite create
#[endpoint(
    post,
    path = "/room/{room_id}/invite",
    tags = ["invite"],
    scopes = [Full],
    permissions_optional = [InviteManage],
    audit_log_events = ["InviteCreate"],
    response(CREATED, body = Invite, description = "success"),
)]
pub mod invite_room_create {
    use crate::v1::types::{Invite, InviteCreate, RoomId};

    pub struct Request {
        #[path]
        pub room_id: RoomId,

        #[json]
        pub invite: InviteCreate,
    }

    pub struct Response {
        #[json]
        pub invite: Invite,
    }
}

/// Invite list room
#[endpoint(
    get,
    path = "/room/{room_id}/invite",
    tags = ["invite"],
    scopes = [Full],
    permissions_optional = [InviteManage],
    response(OK, body = PaginationResponse<Invite>, description = "success"),
)]
pub mod invite_room_list {
    use crate::v1::types::{Invite, InviteCode, PaginationQuery, PaginationResponse, RoomId};

    pub struct Request {
        #[path]
        pub room_id: RoomId,

        #[query]
        pub pagination: PaginationQuery<InviteCode>,
    }

    pub struct Response {
        #[json]
        pub invites: PaginationResponse<Invite>,
    }
}

/// Invite list channel
#[endpoint(
    get,
    path = "/channel/{channel_id}/invite",
    tags = ["invite"],
    scopes = [Full],
    permissions_optional = [InviteManage],
    response(OK, body = PaginationResponse<Invite>, description = "success"),
)]
pub mod invite_channel_list {
    use crate::v1::types::{ChannelId, Invite, InviteCode, PaginationQuery, PaginationResponse};

    pub struct Request {
        #[path]
        pub channel_id: ChannelId,

        #[query]
        pub pagination: PaginationQuery<InviteCode>,
    }

    pub struct Response {
        #[json]
        pub invites: PaginationResponse<Invite>,
    }
}

/// Invite channel create
#[endpoint(
    post,
    path = "/channel/{channel_id}/invite",
    tags = ["invite"],
    scopes = [Full],
    permissions_optional = [InviteManage],
    audit_log_events = ["InviteCreate"],
    response(CREATED, body = Invite, description = "success"),
)]
pub mod invite_channel_create {
    use crate::v1::types::{ChannelId, Invite, InviteCreate};

    pub struct Request {
        #[path]
        pub channel_id: ChannelId,

        #[json]
        pub invite: InviteCreate,
    }

    pub struct Response {
        #[json]
        pub invite: Invite,
    }
}

/// Invite update
#[endpoint(
    patch,
    path = "/invite/{invite_code}",
    tags = ["invite"],
    scopes = [Full],
    permissions_optional = [InviteManage],
    audit_log_events = ["InviteUpdate"],
    response(OK, body = Invite, description = "success"),
)]
pub mod invite_update {
    use crate::v1::types::{Invite, InviteCode, InvitePatch};

    pub struct Request {
        #[path]
        pub invite_code: InviteCode,

        #[json]
        pub patch: InvitePatch,
    }

    pub struct Response {
        #[json]
        pub invite: Invite,
    }
}

/// Invite use
#[endpoint(
    post,
    path = "/invite/{invite_code}",
    tags = ["invite"],
    scopes = [Full],
    response(OK, description = "success"),
)]
pub mod invite_use {
    use crate::v1::types::InviteCode;

    pub struct Request {
        #[path]
        pub invite_code: InviteCode,
    }

    pub struct Response {}
}

/// Invite server create
#[endpoint(
    post,
    path = "/server/invite",
    tags = ["invite"],
    scopes = [Full],
    permissions_optional = [InviteCreate],
    response(CREATED, body = Invite, description = "success"),
)]
pub mod invite_server_create {
    use crate::v1::types::{Invite, InviteCreate};

    pub struct Request {
        #[json]
        pub invite: InviteCreate,
    }

    pub struct Response {
        #[json]
        pub invite: Invite,
    }
}

/// Invite server list
#[endpoint(
    get,
    path = "/server/invite",
    tags = ["invite"],
    scopes = [Full],
    permissions_optional = [InviteManage],
    response(OK, body = PaginationResponse<Invite>, description = "success"),
)]
pub mod invite_server_list {
    use crate::v1::types::{Invite, InviteCode, PaginationQuery, PaginationResponse};

    pub struct Request {
        #[query]
        pub pagination: PaginationQuery<InviteCode>,
    }

    pub struct Response {
        #[json]
        pub invites: PaginationResponse<Invite>,
    }
}

/// Invite user create
#[endpoint(
    post,
    path = "/user/{user_id}/invite",
    tags = ["invite"],
    scopes = [Full],
    permissions_optional = [InviteCreate],
    response(CREATED, body = Invite, description = "success"),
)]
pub mod invite_user_create {
    use crate::v1::types::misc::UserIdReq;
    use crate::v1::types::{Invite, InviteCreate};

    pub struct Request {
        #[path]
        pub user_id: UserIdReq,

        #[json]
        pub invite: InviteCreate,
    }

    pub struct Response {
        #[json]
        pub invite: Invite,
    }
}

/// Invite user list
#[endpoint(
    get,
    path = "/user/{user_id}/invite",
    tags = ["invite"],
    scopes = [Full],
    response(OK, body = PaginationResponse<Invite>, description = "success"),
)]
pub mod invite_user_list {
    use crate::v1::types::misc::UserIdReq;
    use crate::v1::types::{Invite, InviteCode, PaginationQuery, PaginationResponse};

    pub struct Request {
        #[path]
        pub user_id: UserIdReq,

        #[query]
        pub pagination: PaginationQuery<InviteCode>,
    }

    pub struct Response {
        #[json]
        pub invites: PaginationResponse<Invite>,
    }
}
